File: test/Systems_With_Delays/CV_Model.jl
module CV_Model

using DifferentialEquations
using CSV, DataFrames
using Plots
using Dates

const ABS_TOL = 1e-9
const REL_TOL = 1e-9

# Cardiovascular Model
struct Params
    ca::Float64
    cv::Float64
    R::Float64
    r::Float64
    Vstr::Float64
    alpha0::Float64
    alphas::Float64
    alphap::Float64
    alphaH::Float64
    beta0::Float64
    betas::Float64
    betap::Float64
    betaH::Float64
    gammaH::Float64
end

const params = Params(1.55, 519, 1.05, 0.068, 67.9, 93, 93, 93, 0.84, 7, 7, 7, 1.17, 0)
const tau = 4.0

@inline function ddefun(du::Vector{Float64}, u::Vector{Float64}, h, p::Params, t::Float64)
    R = t <= 600 ? 1.05 : 0.21 * exp(600 - t) + 0.84
    ylag = h(p, t - tau)
    Patau = ylag[1]
    Paoft = u[1]
    Pvoft = u[2]
    Hoft = u[3]

    du[1] = - (1 / (p.ca * R)) * Paoft + (1 / (p.ca * R)) * Pvoft + (1 / p.ca) * p.Vstr * Hoft
    du[2] = (1 / (p.cv * R)) * Paoft - (1 / (p.cv * R) + 1 / (p.cv * p.r)) * Pvoft
    Ts = 1 / (1 + (Patau / p.alphas)^p.betas)
    Tp = 1 / (1 + (p.alphap / Paoft)^p.betap)
    du[3] = (p.alphaH * Ts) / (1 + p.gammaH * Tp) - p.betaH * Tp
end

@inline function history(p::Params, t::Float64)
    P0 = 93
    Paval = P0
    Pvval = (1 / (1 + p.R / p.r)) * P0
    Hval = (1 / (p.R * p.Vstr)) * (1 / (1 + p.r / p.R)) * P0
    return [Paval, Pvval, Hval]
end

const u0 = [93, (1 / (1 + params.R / params.r)) * 93, (1 / (params.R * params.Vstr)) * (1 / (1 + params.r / params.R)) * 93]
const tspan = (0.0, 1000.0)

function Run_CV_Model()
    prob = DDEProblem(ddefun, u0, history, tspan, params; constant_lags = [tau], tstops = [600])
    alg = MethodOfSteps(Tsit5())

    println("Solving Cardiovascular model...")
    start = Dates.now()
    sol = solve(prob, alg, reltol=REL_TOL, abstol=ABS_TOL)
    elapsed_time = Dates.now() - start
    println("It took ", elapsed_time, " seconds to solve the Cardiovascular model")
    
    t_interp = range(0, stop=1000, length=1000)
    sol_interp = sol(t_interp)
    
    data = DataFrame(time=t_interp, Pa=sol_interp[1,:], Pv=sol_interp[2,:], HR=sol_interp[3,:])
    CSV.write("data/cardiovascular_model_jl.csv", data)
    
    println("Cardiovascular model solved successfully")
    return elapsed_time
end

end # module

File: comparison_tests/dde_tests_compiled/Cardiovascular_Model/cv_model.c
#include <dde/dde.h>
#include <stddef.h>
#include <math.h>

// Parameters (passed in order):
//   ca, cv, R, r, Vstr, alpha0, alphas, alphap, alphaH,
//   beta0, betas, betap, betaH, gammaH, tau
void cv_model(size_t n, double t, double *y, double *dydt, void *data) {
  double *pars   = (double*) data;
  double ca      = pars[0];
  double cv_     = pars[1];
  // pars[2] is R0 (unused; R is computed dynamically below)
  double r       = pars[3];
  double Vstr    = pars[4];
  double alphas  = pars[6];
  double alphap  = pars[7];
  double alphaH  = pars[8];
  double betas   = pars[10];
  double betap   = pars[11];
  double betaH   = pars[12];
  double gammaH  = pars[13];
  double tau     = pars[14];

  // Time-varying resistance (matches R model exactly)
  double R = (t <= 600.0) ? 1.05 : 0.21 * exp(600.0 - t) + 0.84;

  // Lag: y[0] (Pa, 0-indexed) at t - tau
  static const int idx[1] = {0};
  double Patau[1];
  ylag_vec_int(t - tau, idx, 1, Patau);

  double Pa = y[0];
  double Pv = y[1];
  double H  = y[2];

  double dydt0 = -(1.0 / (ca * R)) * Pa
               +  (1.0 / (ca * R)) * Pv
               +  (1.0 / ca) * Vstr * H;

  double dydt1 =  (1.0 / (cv_ * R)) * Pa
               - ((1.0 / (cv_ * R)) + (1.0 / (cv_ * r))) * Pv;

  double Ts   = 1.0 / (1.0 + pow(Patau[0] / alphas, betas));
  double Tp   = 1.0 / (1.0 + pow(alphap / Pa,       betap));
  double dydt2 = (alphaH * Ts) / (1.0 + gammaH * Tp) - betaH * Tp;

  dydt[0] = dydt0;
  dydt[1] = dydt1;
  dydt[2] = dydt2;
}

#include <dde/dde.c>